llms/mlx_lm/tuner/trainer.py

```diff
@@ -118,6 +118,28 @@ def evaluate(
     return np.sum(all_losses) / ntokens
 
 
+class TrainingCallback:
+    def on_train_begin(self, iters):
+        """Called at the beginning of training."""
+        pass
+
+    def on_train_loss_report(self, steps: int, loss: float, it_sec: float, tokens_sec: float):
+        """Called to report training loss at specified intervals."""
+        pass
+
+    def on_val_loss_report(self, steps: int, loss: float, took: float):
+        """Called to report validation loss at specified intervals or the beginning."""
+        pass
+
+    def on_checkpoint_save(self, steps: int, file: str):
+        """Called to save model checkpoints at specified intervals."""
+        pass
+
+    def on_train_end(self, file: str):
+        """Called at the end of training."""
+        pass
```

Comment:
I'm wondering if we need all of this or if it could be simplified to just a single function? Like for the example you gave logging to wanddb, just having a little callback which saves train and validation loss would work right? Or did you need the callbacks in other places as well?

Reply:
There is no strong necessity. You can also know the start and end by calling the train function itself. I removed this part in the new submission.

But if the callback contains these two functions, more complete logic can be executed in the callback. At the same time, there is no need to put some initialization code logic together with the logic of calling the train function. Of course, this flexibility will increase the complexity of the code.